File: src/algorithms/01_sorting/03_insertion_sort/implementations/go/insertion_sort.go
package insertionSort

func InsertionSort(unsortedArray []int) []int {
	clonedArray := unsortedArray

	for i := 1; i < len(unsortedArray); i++ {
		key := clonedArray[i]
		j := i

		for j > 0 && clonedArray[j-1] > key {
			clonedArray[j] = clonedArray[j-1]
			j--
		}

		clonedArray[j] = key
	}

	return clonedArray
}

File: src/algorithms/01_sorting/03_insertion_sort/implementations/ts/insertion_sort.ts
export const insertion_sort = (unsorted_array: Array<number>): Array<number> => {
  const cloned_array = [...unsorted_array]

  // Iterate from second item to last
  for (let i = 1; i < cloned_array.length; i++) {
    const key = cloned_array[i]; // Value of current key position, start in second element
    let j = i;

    while (j > 0 && cloned_array[j - 1] > key) {
      cloned_array[j] = cloned_array[j - 1]; // Replace second element with first one
      j--; // Decrease counter, so we perform comparison and replace previous element if necessary
    }

    cloned_array[j] = key; // End of outer loop, insert key
  }

  return cloned_array
};

